### PR TITLE
Add wifi support

### PR DESCRIPTION
**Pull request by Jason-ZW**
### Features
- the configurable setting without requiring the reboot
- both support static and dynamic IP
### Limitations
- multiple wifi is not supported on the same network interface
### Usage

```
# configure cloud-config.yml
rancher:
  network:
    interfaces:
      wlan0:
        wifi_network: network1
      wlan1:
        wifi_network: network2
      wlan2:
        wifi_network: network3
    wifi_networks:
      network1:
        ssid: RancherLabs_SY
        psk: xxxxxxxxxxx
        scan_ssid: 1
        address: 192.168.1.79/24
        gateway: 192.168.1.1
      network2:
        ssid: RancherLabs_SY
        psk: xxxxx
        scan_ssid: 1
        address: 192.168.1.80/24
        gateway: 192.168.1.1
      network3:
         ssid: RancherLabs_Test
         psk: xxxxx
         scan_ssid: 1
         key_mgmt: NONE

# restart network container
$ system-docker restart network
```

### CODE REVIEW

File: cmd/cloudinitsave/cloudinitsave.go [hunk 1]
```diff
@@ -66,8 +66,10 @@ func Main() {
 		log.Errorf("Failed to save cloud-config: %v", err)
 	}
 
+	// exit wpa_supplicant
+	netconf.StopWpaSupplicant()
```
Comment (niusmallnan) [inline cmd/cloudinitsave/cloudinitsave.go]: Not sure but if we don't set any WiFi networks, is there a problem with forcibly stopping WPA?

File: pkg/netconf/netconf_linux.go [hunk 2]
```diff
@@ -208,6 +211,10 @@ func applyOuter(link netlink.Link, netCfg *NetworkConfig, wg *sync.WaitGroup, us
 	}
 
 	log.Debugf("Config(%s): %#v", linkName, match)
+	if match.WifiNetwork != "" {
+		match.DHCP = true
```
Comment (niusmallnan) [inline pkg/netconf/netconf_linux.go]: I think we should add some comments here, it is necessary to explain why we set dhcp true.
```
We plan to use the dhcpcd hook to control the wpa_supplicant, https://wiki.archlinux.org/index.php/Dhcpcd#Hooks. Whether the Wi-Fi network uses DHCP or Static mode, the dhcpcd should be enabled by default, So we set `DHCP=true` here.
```

File: config/types.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ const (
 	MetaDataFile           = "/var/lib/rancher/conf/metadata"
 	CloudConfigFile        = "/var/lib/rancher/conf/cloud-config.yml"
 	EtcResolvConfFile      = "/etc/resolv.conf"
+	WPAFile                = "/etc/wpa_supplicant-%s.conf"
```
[the remark this answers is not in the document]
Comment (Jason-ZW) [reply config/types.go]: Already modified

File: pkg/netconf/netconf_linux.go [hunk 4]
```diff
@@ -530,11 +564,23 @@ func GetValidLinkList() ([]netlink.Link, error) {
 	return validLinkList, nil
 }
 
-func StopDhcpcd() {
+func StopDhcpcd(iface string) {
+	if iface != "" {
+		exitDhcpArgs = append(exitDhcpArgs, iface)
```
[the remark this answers is not in the document]
Comment (Jason-ZW) [reply pkg/netconf/netconf_linux.go]: Already modified

File: pkg/netconf/netconf_linux.go [hunk 5]
```diff
@@ -530,11 +564,23 @@ func GetValidLinkList() ([]netlink.Link, error) {
 	return validLinkList, nil
 }
 
-func StopDhcpcd() {
+func StopDhcpcd(iface string) {
```
[the remark this answers is not in the document]
Comment (Jason-ZW) [reply pkg/netconf/netconf_linux.go]: Already modified

File: cmd/cloudinitsave/cloudinitsave.go (hunk 1 above)
Comment (Jason-ZW) [reply cmd/cloudinitsave/cloudinitsave.go]: Only print log on network container: `Failed to run cmd [wpa_cli terminate]: exist status 255`

File: cmd/network/network_templates.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
+package network
+
+import (
+	"strconv"
+	"text/template"
+)
+
+const wpaTemplate = `
```
Comment (niusmallnan) [inline cmd/network/network_templates.go]: Is it possible that we put followings into a tpl file？
https://github.com/rancher/os/blob/master/images/01-base/Dockerfile

Then we can read  this file as template, execute it and write it back.

File: cmd/network/network_templates.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	private_key_passwd="{{$v}}"
+	{{- else}}
+	private_key{{addFunc $i 1}}_passwd="{{$v}}"
+	{{- end}}
+	{{- end}}
+}			
+{{- end}}
+`
+
+const dhcpTemplate = `
```
Comment (niusmallnan) [inline cmd/network/network_templates.go]: Same suggestion with wpaTemplate.

File: cmd/cloudinitsave/cloudinitsave.go (hunk 1 above)
Comment (niusmallnan) [reply cmd/cloudinitsave/cloudinitsave.go]: It's ok.

File: cmd/network/network.go [hunk 8]
```diff
@@ -86,3 +95,46 @@ func ApplyNetworkConfig(cfg *config.CloudConfig) {
...
+				log.Errorf("Failed to open file: %s err: %v", filename, err)
+			}
+			if c, ok := networks[v.WifiNetwork]; ok {
+				configs[v.WifiNetwork] = c
+			}
+			p := template.Must(template.New(fmt.Sprintf("wpa-%s", k)).Funcs(funcMap).Parse(wpaTemplate))
+			if err = p.Execute(f, configs); err != nil {
+				log.Errorf("Failed to wrote wpa configuration to %s: %v", filename, err)
+			}
+			f.Close()
```
[the remark this answers is not in the document]
Comment (Jason-ZW) [reply cmd/network/network.go]: Already modified

File: cmd/network/network.go [hunk 9]
```diff
@@ -51,6 +54,12 @@ func ApplyNetworkConfig(cfg *config.CloudConfig) {
 	if cfg.Rancher.Network.DHCPTimeout <= 0 {
 		cfg.Rancher.Network.DHCPTimeout = cfg.Rancher.Defaults.Network.DHCPTimeout
 	}
+
+	if len(cfg.Rancher.Network.WifiNetworks) > 0 {
+		generateDhcpcdFiles(cfg)
```
[the remark this answers is not in the document]
Comment (Jason-ZW) [reply cmd/network/network.go]: Already modified

**Pull request merged**